src/libcore/ops/drop.rs

```diff
@@ -1,85 +1,136 @@
+/// please see [the reference] section on destructors.
 ///
-/// Because of this recursive dropping, you do not need to implement this trait
-/// unless your type needs its own destructor logic.
+/// [the reference]: https://doc.rust-lang.org/reference/destructors.html
 ///
-/// Refer to [the chapter on `Drop` in *The Rust Programming Language*][book]
-/// for some more elaboration.
+/// The `Drop` trait provides a way to customize the destructor of a type.
+/// While this is not necessary for most types, there are some cases where
+/// it is useful. Most notably when dealing with resources. That resource may
+/// be memory, it may be a file descriptor, it may be a network socket.
+/// But when the type is no longer going to be used, it should "clean up" that
+/// resource by freeing the memory or closing the file or socket. This is
+/// the job of a destructor, and therefore the job of `Drop::drop`.
 ///
-/// [book]: ../../book/ch15-03-drop.html
+/// ## Using the `Drop` trait
 ///
-/// # Examples
+/// When a value goes out of scope, the compiler automatically inserts "drop glue"
+/// for this value. This means calling to `Drop::drop` for it, and then recursively for all
+/// other values it contains:
 ///
-/// ## Implementing `Drop`
-///
-/// The `drop` method is called when `_x` goes out of scope, and therefore
-/// `main` prints `Dropping!`.
-///
-/// ```
+/// ```rust
 /// struct HasDrop;
 ///
 /// impl Drop for HasDrop {
 ///     fn drop(&mut self) {
-///         println!("Dropping!");
+///         println!("Dropping HasDrop!");
+///     }
+/// }
+///
+/// struct HasTwoDrops {
+///     one: HasDrop,
+///     two: HasDrop,
+/// }
+///
+/// impl Drop for HasTwoDrops {
+///     fn drop(&mut self) {
+///         println!("Dropping HasTwoDrops!");
 ///     }
 /// }
 ///
 /// fn main() {
-///     let _x = HasDrop;
+///     let _x = HasTwoDrops { one: HasDrop, two: HasDrop };
+///     println!("Running!");
 /// }
 /// ```
 ///
-/// ## Dropping is done recursively
+/// Running this will print
+///
+/// ```text
+/// Running!
+/// Dropping HasTwoDrops!
+/// Dropping HasDrop!
+/// Dropping HasDrop!
+/// ```
 ///
-/// When `outer` goes out of scope, the `drop` method will be called first for
-/// `Outer`, then for `Inner`. Therefore, `main` prints `Dropping Outer!` and
-/// then `Dropping Inner!`.
+/// Even if we remove the implementation of `Drop` for `HasTwoDrop`, the destructors of its fields are still called.
+/// This would result in
 ///
+/// ```test
+/// Running!
+/// Dropping HasDrop!
+/// Dropping HasDrop!
 /// ```
-/// struct Inner;
-/// struct Outer(Inner);
 ///
-/// impl Drop for Inner {
+/// ## You cannot call `Drop::drop` yourself
+///
+/// Because the compiler automatically calls to `Drop::drop`, you cannot call it yourself. This
+/// would lead to a "double drop", where `Drop::drop` is called twice on the same value. This
+/// can lead to things like "double frees".
```

Comment:
That argument is... a bit strange to me. I can call `mem::drop` myself without double-dropping, so if that was the problem, why wouldn't the compiler just apply the same rules for `Drop::drop`?

The answer is that *the type of `Drop::drop` is a lie*. That function is *not* like any other function that takes a mutable reference. Usually, a function taking `&mut T` borrows `T` and then it has to give back an intact `T` when the borrow is over. `Drop::drop`, on the other hand, partially *destroys* the `T`, and it may rely on the fact that no other method is called on this type any more after it returns. *That* is why it cannot be called directly.

Reply:
:+1:

Reply:
improved